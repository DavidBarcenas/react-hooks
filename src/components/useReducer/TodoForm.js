import React from 'react'
import { useForm } from '../../hooks/useForm'

export const TodoForm = ({handleAddTodo}) => {

  const [{desc}, handleInputchange, reset] = useForm({ desc: '' })

  const handleSubmit = (e) => {
    e.preventDefault();

    if(desc.trim().length <= 1) { return; }

    const newTodo = { id: new Date().getTime(), desc, done: false }
    
    handleAddTodo(newTodo);
    reset();
  }

  return (
    <form className="mb-3" onSubmit={handleSubmit}>
      <input type="text" name="desc" placeholder="ToDo" value={desc} onChange={handleInputchange}/>
      <button type="submit" className="btn btn-success btn-sm mb-1">Add ToDo</button>
    </form>
  )
}
