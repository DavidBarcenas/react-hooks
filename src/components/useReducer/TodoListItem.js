import React from 'react'
import PropTypes from 'prop-types'

export const TodoListItem = ({todo, i, handleToogle, handleDelete }) => {
  return (
    <li className="list-group-item d-flex justify-content-between align-items-center">
      <label className={`${todo.done && 'complete'} mb-0`} onClick={() => handleToogle(todo.id)}>{i + 1} - {todo.desc}</label>
      <button className="btn btn-danger btn-sm" onClick={() => handleDelete(todo.id)}>X</button>
    </li>
  )
}

TodoListItem.propTypes = {
  todo: PropTypes.object.isRequired,
  i: PropTypes.number.isRequired,
  handleToogle: PropTypes.func.isRequired,
  handleDelete: PropTypes.func.isRequired,
}