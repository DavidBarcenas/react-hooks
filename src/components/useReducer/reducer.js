const initialState = [{
  id: 1,
  todo: 'Create app',
  done: false
}]


const todoReducer = (state = initialState, action) => {

  if(action.type=== 'agregar') {
    return [...state, action.payload]
  }

  return state;
}

let todos;

const newTodo = {
  id: 2,
  todo: 'Add wallet',
  done: false
}

const action = {
  type: 'agregar',
  payload: newTodo
}

todos = todoReducer(todos, action)


console.log(todos)