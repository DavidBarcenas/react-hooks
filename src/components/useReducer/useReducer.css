.list-group-item {
  cursor: pointer;
}

.complete {
  text-decoration: line-through;
}