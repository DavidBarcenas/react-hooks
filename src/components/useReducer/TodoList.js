import React from 'react'
import PropTypes from 'prop-types';
import { TodoListItem } from './TodoListItem'

export const TodoList = ({todos, handleToogle, handleDelete}) => {
  return (
    <ul className="list-group text-left">
      {
        todos.map((todo, i) => 
          <TodoListItem 
            key={todo.id} 
            todo={todo} 
            i={i} 
            handleToogle={handleToogle} 
            handleDelete={handleDelete} />
          )
      }
    </ul>
  )
}

TodoList.propTypes = {
  todos: PropTypes.array.isRequired,
  handleToogle: PropTypes.func.isRequired,
  handleDelete: PropTypes.func.isRequired,
}