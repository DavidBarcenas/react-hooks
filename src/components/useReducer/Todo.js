import React, { useReducer, useEffect } from 'react'
import { todoReducer } from './todoReducer'
import { TodoList } from './TodoList'
import { TodoForm } from './TodoForm'
import './useReducer.css'

// const initialState = [{ id: new Date().getTime(), desc: 'Learning useReducer', done: false }]
const init = () => JSON.parse(localStorage.getItem('todos')) || [];

export const Todo = () => {

  const [todos, dispatch] = useReducer(todoReducer, [], init)

  useEffect(() => {
    localStorage.setItem('todos', JSON.stringify(todos))
  }, [todos])

  const handleDelete  = (todoID) => dispatch({ type: 'delete', payload: todoID });
  const handleToogle  = (todoID) => dispatch({ type: 'toggle', payload: todoID });
  const handleAddTodo = (newTodo) => dispatch({ type: 'add', payload: newTodo });

  return (
    <div className="text-center p-5">
      <h1 className="text-primary">Todo App ({todos.length})</h1>
      <hr/>

      <TodoForm handleAddTodo={handleAddTodo} />

      <TodoList 
        todos={todos}
        handleToogle={handleToogle}
        handleDelete={handleDelete} 
      />
    </div>
  )
}
