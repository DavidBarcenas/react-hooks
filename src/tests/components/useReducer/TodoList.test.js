import '@testing-library/jest-dom';
import React from 'react'
import { shallow } from 'enzyme';
import { demoTodos } from '../../fixtures/demoTodos';
import { TodoList } from '../../../components/useReducer/TodoList';

describe('<TodoList /> testing', () => {
  
  const wrapper = shallow(
    <TodoList 
      todos={demoTodos}
      handleToogle={jest.fn()}
      handleDelete={jest.fn()}
    />
  )


  test('should render correctly', () => {
    expect(wrapper).toMatchSnapshot()
  })
  
  
  test('should have two listItems', () => {
    expect(wrapper.find('TodoListItem').length).toBe(demoTodos.length)
    expect(wrapper.find('TodoListItem').at(0).prop('handleDelete')).toEqual(expect.any(Function))
    expect(wrapper.find('TodoListItem').at(0).prop('handleToogle')).toEqual(expect.any(Function))
  })
  
})
