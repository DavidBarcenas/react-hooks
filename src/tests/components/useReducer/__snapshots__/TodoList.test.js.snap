// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoList /> testing should render correctly 1`] = `
<ul
  className="list-group text-left"
>
  <TodoListItem
    handleDelete={[MockFunction]}
    handleToogle={[MockFunction]}
    i={0}
    key="1"
    todo={
      Object {
        "desc": "Learning React",
        "done": true,
        "id": 1,
      }
    }
  />
  <TodoListItem
    handleDelete={[MockFunction]}
    handleToogle={[MockFunction]}
    i={1}
    key="2"
    todo={
      Object {
        "desc": "Learning Mongo",
        "done": false,
        "id": 2,
      }
    }
  />
</ul>
`;
